src/Features/Core/Portable/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,337 @@
+                // 3. Operation action for field initializers to detect non-constant initialization.
+                // 4. Operation action for invalid operations to bail out on erroneous code.
+                // 5. A symbol start/end action for named types to report diagnostics for candidate members that have no usage in executable code.
+                //
+                // Note that we need to register separately for OperationKind.Invocation due to https://github.com/dotnet/roslyn/issues/26206
+
+                compilationStartContext.RegisterSymbolAction(compilationAnalyzer.AnalyzeSymbolDeclaration, SymbolKind.Method, SymbolKind.Field, SymbolKind.Property, SymbolKind.Event);
+
+                compilationStartContext.RegisterSymbolStartAction(symbolStartContext =>
+                {
+                    var hasInvalidOperation = false;
+                    symbolStartContext.RegisterOperationAction(compilationAnalyzer.AnalyzeMemberReferenceOperation, OperationKind.FieldReference, OperationKind.MethodReference, OperationKind.PropertyReference, OperationKind.EventReference);
+                    symbolStartContext.RegisterOperationAction(compilationAnalyzer.AnalyzeFieldInitializer, OperationKind.FieldInitializer);
+                    symbolStartContext.RegisterOperationAction(compilationAnalyzer.AnalyzeInvocationOperation, OperationKind.Invocation);
+                    symbolStartContext.RegisterOperationAction(_ => hasInvalidOperation = true, OperationKind.Invalid);
+                    symbolStartContext.RegisterSymbolEndAction(symbolEndContext => compilationAnalyzer.OnSymbolEnd(symbolEndContext, hasInvalidOperation));
+                }, SymbolKind.NamedType);
+            });
+        }
+
+        private sealed class CompilationAnalyzer
+        {
+            private readonly DiagnosticDescriptor _removeUnusedMembersRule, _removeUnreadMembersRule;
+            private readonly object _gate;
+            private readonly Dictionary<ISymbol, ValueUsageInfo> _symbolValueUsageStateMap;
+            private readonly Lazy<INamedTypeSymbol> _lazyTaskType, _lazyGenericTaskType;
+
+            public CompilationAnalyzer(Compilation compilation, DiagnosticDescriptor removeUnusedMembersRule, DiagnosticDescriptor removeUnreadMembersRule)
+            {
+                _removeUnusedMembersRule = removeUnusedMembersRule;
+                _removeUnreadMembersRule = removeUnreadMembersRule;
+                _gate = new object();
+
+                // State map for candidate member symbols, with the value indicating how each symbol is used in executable code.
+                _symbolValueUsageStateMap = new Dictionary<ISymbol, ValueUsageInfo>();
+
+                _lazyTaskType = new Lazy<INamedTypeSymbol>(() => compilation.TaskType());
+                _lazyGenericTaskType = new Lazy<INamedTypeSymbol>(() => compilation.TaskOfTType());
```

Comment:
1. being lazy here is unlikely to be a real savings.  Indeed, IIRC, compilation, and the assembly layers cache GetTypeByMetadataName since it is called so often.
2. is your code resilient to Task or `Task<T>` not being found?  They're not always available in all projects.

Reply:
Made non-lazy. Yes, null values are handled fine.

---
In reply to: [213414786](https://github.com/dotnet/roslyn/pull/29511#discussion_r213414786) [](ancestors = 213414786)